frontend/src/app/utils.ts

```diff
@@ -10,4 +17,29 @@ export function sortBy<T>(array: Array<T>, extractor: (T) => any) {
     }
     return 0;
   });
+  return result;
+}
+
+/**
+ * Takes a string containing placeholders such as ${foo}, and replaces them by their value in the given
+ * parameters. Beware: no expression supported, no space in curly braces.
+ *
+ * So interpolate('hello ${name}', { name: 'JB' }) returns "hello JB".
+ *
+ * This function is only used (currently) to generate functional error messages from code and parameters
+ * coming from the backend.
+ */
+export function interpolate(template: string, parameters: {[key: string]: any}): string {
+  // ugly loop because JS doesn't have a Regexp.quote() method, nor a replaceAll method. replace is supposed to replace
+  // all, but it does not.
+  let result = template;
```

Comment:
That's what I would love to do, but JS always is half-baked:

 - you can pass a g flag if you pass a regex as first argument, and not a string
 - transforming an arbitrary string into a regex is hard, because there is no quote method to escape the special characters
 
lodash has a complex template() function that allows doing that (with arbitrary expressions, etc.), but I thought that given what we need to do, what I did is just fine. It's probably not very efficient, but it will be used very seldomly, and on short strings with very few parameters, so it's OK.